### ISSUE TITLE
Docs need clarification on what "pure" actually means, especially with regards to memory

### ISSUE THREAD

**Issue by alexrp** [Vector35/binaryninja-api#5209]
The docs currently state:

```
Pure. Functions that are pure are assumed to have no side effects and may be inlined if their return value is trivial
```

This misses a lot of nuance:

* Is a function that reads from a constant data variable pure?
* Is a function that reads indirectly from a pointer argument pure?
* Is a function that writes indirectly to a pointer argument pure?

etc. Right now, I can only safely assume that functions can be marked pure if they never perform memory access.

Could refer to the docs for GCC's [`pure`](https://gcc.gnu.org/onlinedocs/gcc/Common-Function-Attributes.html#index-pure-function-attribute) and [`const`](https://gcc.gnu.org/onlinedocs/gcc/Common-Function-Attributes.html#index-const-function-attribute) function attributes when writing these docs as I suspect many C/C++ developers will be somewhat familiar with these.

**Comment by alexrp** [comment]
It might also be good to mention if there are any extra implications of applying `__pure` to a function pointer. (I expect not?)

**Comment by CouleeApps** [comment]
This is included in >= 4.1.5339-dev
